### PR TITLE
Feature/type inspection

### PR DESCRIPTION
**Pull request by mpoeter**
Add type inspection library for easy serialization/deserialization of custom structs.

This library is heavily inspired by the type inspection framework of the C++ Actor Framework (CAF): https://github.com/actor-framework/actor-framework/

### CODE REVIEW

File: include/velocypack/inspection/Inspector.h [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    static constexpr const char InvariantFailedError[] =
+        "Object invariant failed";
+
+   private:
+    template<class TT>
+    friend struct Object;
+    FieldsResult(Result&& res, T& object)
+        : result(std::move(res)), object(object) {}
+    Result result;
```
Comment (cpjulia) [inline include/velocypack/inspection/Inspector.h:87]: _result as is member?

File: include/velocypack/inspection/Inspector.h [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+  template<class T>
+  [[nodiscard]] Result apply(T& x) {
+    return process(self(), x);
+  }
+
+  template<class T>
+  struct Object;
+
+  template<class T, const char ErrorMsg[], class Func, class... Args>
+  static Result checkInvariant(Func&& func, Args&&... args) {
```
Comment (cpjulia) [inline include/velocypack/inspection/Inspector.h:48]: can this function be a constexpr? Like 
```suggestion
  static constexpr Result checkInvariant(Func&& func, Args&&... args) {
```

File: include/velocypack/inspection/Inspector.h [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+                    "Invariants must either return bool or "
+                    "velocypack::inspection::Result");
+      return std::invoke(std::forward<Func>(func), std::forward<Args>(args)...);
+    }
+  }
+
+  template<class T>
+  struct FieldsResult {
+    template<class Invariant>
+    Result invariant(Invariant&& func) {
```
Comment (cpjulia) [inline include/velocypack/inspection/Inspector.h:66]: can this one be constexpr as well?

File: include/velocypack/inspection/Inspector.h (hunk 2 above)
Comment (cpjulia) [reply include/velocypack/inspection/Inspector.h:48]: maybe is a candidate for
```suggestion
  [[nodiscard]] static constexpr Result checkInvariant(Func&& func, Args&&... args) {
```
?

File: include/velocypack/inspection/Inspector.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+      return {inspector.endObject(), object};
+    }
+
+   private:
+    friend struct InspectorBase;
+    explicit Object(Derived& inspector, T& o)
+        : inspector(inspector), object(o) {}
+
+    Derived& inspector;
```
Comment (cpjulia) [inline include/velocypack/inspection/Inspector.h:112]: Maybe this and other members should start with a `_`?

**Review by maierlars — APPROVED** [review]

File: include/velocypack/inspection/InspectorAccess.h [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+
+template<class T, class Inspector>
+constexpr inline bool IsInspectable() {
+  return IsBuiltinType<T>() || HasInspectOverload<T, Inspector>::value ||
+         IsListLike<T>::value || IsMapLike<T>::value || IsTuple<T>::value ||
+         HasInspectorAccessSpecialization<T>();
+}
+
+template<class Inspector, class T>
+[[nodiscard]] Result process(Inspector& f, T& x) {
```
Comment (cpjulia) [inline include/velocypack/inspection/InspectorAccess.h:178]: can be constexpr method?

File: include/velocypack/inspection/LoadInspector.h [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+      return res;
+    }
+    for (auto&& pair : VPackObjectIterator(_slice)) {
+      LoadInspector ff(pair.value, _options);
+      typename T::mapped_type val;
+      if (auto res = process(ff, val); !res.ok()) {
+        return {std::move(res), "'" + pair.key.copyString() + "'",
+                Result::ArrayTag{}};
+      }
+      map.emplace(pair.key.copyString(), std::move(val));
```
Comment (cpjulia) [inline include/velocypack/inspection/LoadInspector.h:134]: Is is always guaranteed that T is a type that is "emplaceable"? I mean, can it never go wrong to be called with a type that you can't call emplace to? And can it be overwitten (try_emplace)?

**Review by cpjulia — APPROVED** [review]
LGTM

**Comment by mpoeter** [comment]
Closed in favor of PR in ArangoDB: https://github.com/arangodb/arangodb/pull/15944

**Pull request closed without merging**